### PR TITLE
Another layer of indirection to defer Conscrypt init

### PR DESCRIPTION
**Pull request by yschimke**
Avoid https://github.com/square/okhttp/issues/5760, by deferring work from building the platform until when it first encounters an unknown socket type.

This is high priority because the log warning causes a lot of bad stackoverflow advice to add conscrypt dependencies, which is not the intention. https://stackoverflow.com/search?q=org.conscrypt.ConscryptHostnameVerifier

### CODE REVIEW

File: okhttp/src/main/kotlin/okhttp3/internal/platform/android/AndroidSocketAdapter.kt [hunk 1]
```diff
@@ -97,11 +95,51 @@ open class AndroidSocketAdapter(private val sslSocketClass: Class<in SSLSocket>)
...
+        buildIfSupported(sslSocketClass)
       } catch (e: Exception) {
-        Platform.get().log(level = Platform.WARN, message = "unable to load android socket classes", t = e)
+        Platform.get()
+            .log(level = Platform.WARN, message = "unable to load android socket classes", t = e)
         null
       }
     }
+
+    fun buildIfSupported(actualSSLSocketClass: Class<in SSLSocket>): SocketAdapter? {
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/internal/platform/android/AndroidSocketAdapter.kt:106]: Docs on what the expected inputs are please

File: okhttp/src/main/kotlin/okhttp3/internal/platform/android/DeferredSocketAdapter.kt [hunk 2]
```diff
@@ -16,25 +16,25 @@
...
+class DeferredSocketAdapter(private val socketAdapterFactory: Factory) : SocketAdapter {
   private var delegate: SocketAdapter? = null
 
   override fun isSupported(): Boolean {
     return true
   }
 
   override fun matchesSocket(sslSocket: SSLSocket): Boolean {
-    return sslSocket.javaClass.name.startsWith(socketPackage)
+    if (socketAdapterFactory.matchesSocket(sslSocket)) {
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/internal/platform/android/DeferredSocketAdapter.kt:33]: return socketAdapterFactory.matchesSocket(sslSocket)

**Comment by yschimke** [comment]
@swankjesse I'll double back and check these again to be sure with real devices and simulators.

**Comment by yschimke** [comment]
@swankjesse I attempted to squash merge this and it gave me an error with a Web button to retry, that however merged without squashing, I presume a github bug.  But it's exposed some of my transient branch commits onto master.  Sorry.  Check if you want to revert before the next commit and I'll reapply.

I'm not doing that because I don't want to rewrite history etc with a force push.

**Comment by swankjesse** [comment]
I cleaned up our Git history. The updated merge commit is here:
d8f02f4efe2ee98aad80767a6ff4f607784ce831

**Pull request merged**